File: notion/commands/page.py
"""Page commands."""
from __future__ import annotations
import json
import sys

import click

from notion.client import NotionClient, _normalise_id
from notion.errors import NotionCliError, handle_error
from notion.output import print_json, extract_page_properties


def _parse_page_data(data: str | None, properties: tuple[str, ...]) -> dict:
    """Parse --data and KEY=VALUE args into a dict."""
    result = {}
    if data:
        if data == "-":
            raw = sys.stdin.read()
        elif data.startswith("@"):
            raw = open(data[1:]).read()
        else:
            raw = data
        result.update(json.loads(raw))
    for prop in properties:
        if "=" not in prop:
            raise click.BadParameter(f"Expected KEY=VALUE, got: {prop!r}")
        key, _, value = prop.partition("=")
        result[key] = value
    return result


@click.group()
def page():
    """Page commands."""
    pass


@page.command("create")
@click.argument("parent_id")
@click.option("--title", required=True, help="Page title")
@click.option("--data", default=None, help="Additional properties JSON or @file or -")
@click.option("--output", type=click.Choice(["json", "id"]), default="json")
def create(parent_id, title, data, output):
    """Create a new page under a parent page or database."""
    try:
        pid = _normalise_id(parent_id)
        client = NotionClient()

        # Build properties
        extra_props = {}
        if data:
            extra_props = _parse_page_data(data, ())

        # Determine parent type — try as database first, fall back to page
        # Notion API requires specifying parent type
        body = {
            "parent": {"page_id": pid},
            "properties": {
                "title": {
                    "title": [{"type": "text", "text": {"content": title}}]
                }
            }
        }

        if extra_props:
            for key, value in extra_props.items():
                body["properties"][key] = value

        result = client.post("/pages", body)
        client.close()

        if output == "id":
            print(result.get("id", ""))
        else:
            print_json(result)

    except NotionCliError as e:
        handle_error(e)


@page.command("get")
@click.argument("page_id")
@click.option("--output", type=click.Choice(["json", "properties"]), default="json")
def get(page_id, output):
    """Get a page by ID."""
    try:
        pid = _normalise_id(page_id)
        client = NotionClient()
        result = client.get(f"/pages/{pid}")
        client.close()

        if output == "properties":
            print_json(extract_page_properties(result))
        else:
            print_json(result)

    except NotionCliError as e:
        handle_error(e)


@page.command("get-property")
@click.argument("page_id")
@click.argument("property_name")
def get_property(page_id, property_name):
    """Get a specific property value from a page."""
    try:
        pid = _normalise_id(page_id)
        client = NotionClient()
        result = client.get(f"/pages/{pid}")
        client.close()

        props = result.get("properties", {})

        # Case-insensitive search
        found = None
        for name, value in props.items():
            if name == property_name or name.lower() == property_name.lower():
                found = value
                break

        if found is None:
            from notion.errors import NotFoundError
            raise NotFoundError(
                f"Property {property_name!r} not found on page",
                {"available": list(props.keys())}
            )

        print_json(found)

    except NotionCliError as e:
        handle_error(e)


@page.command("set")
@click.argument("page_id")
@click.argument("properties", nargs=-1, metavar="KEY=VALUE")
@click.option("--data", default=None, help="JSON data or @file or -")
def set(page_id, properties, data):
    """Update page properties."""
    try:
        pid = _normalise_id(page_id)
        user_data = _parse_page_data(data, properties)

        if not user_data:
            raise click.UsageError("No properties provided. Use KEY=VALUE args or --data")

        client = NotionClient()

        # Get the page to find parent database (for schema resolution)
        current_page = client.get(f"/pages/{pid}")
        parent = current_page.get("parent", {})

        if parent.get("type") == "database_id":
            from notion.schema import SchemaCache, PropertyResolver
            db_id = parent["database_id"].replace("-", "")
            cache = SchemaCache()
            db_schema = cache.get(db_id)
            if db_schema is None:
                db_result = client.get(f"/databases/{db_id}")
                db_schema = db_result.get("properties", {})
                cache.set(db_id, db_schema)

            resolver = PropertyResolver()
            resolved = resolver.resolve_all(db_schema, user_data)
        else:
            # Standalone page — use raw properties
            resolved = user_data

        result = client.patch(f"/pages/{pid}", {"properties": resolved})
        client.close()
        print_json(result)

    except (click.UsageError,):
        raise
    except NotionCliError as e:
        handle_error(e)


@page.command("append")
@click.argument("page_id")
@click.option("--data", required=True, help="Markdown text, JSON blocks, @file, or -")
def append(page_id, data):
    """Append blocks to a page."""
    try:
        pid = _normalise_id(page_id)

        if data == "-":
            raw = sys.stdin.read()
        elif data.startswith("@"):
            raw = open(data[1:]).read()
        else:
            raw = data

        # Try parsing as JSON first; fall back to markdown
        blocks = None
        try:
            parsed = json.loads(raw)
            if isinstance(parsed, list):
                blocks = parsed
            elif isinstance(parsed, dict) and "children" in parsed:
                blocks = parsed["children"]
        except json.JSONDecodeError:
            pass

        if blocks is None:
            from notion.coerce import markdown_to_blocks
            blocks = markdown_to_blocks(raw)

        client = NotionClient()
        result = client.patch(f"/blocks/{pid}/children", {"children": blocks})
        client.close()
        print_json(result)

    except NotionCliError as e:
        handle_error(e)


@page.command("delete")
@click.argument("page_id")
@click.option("--confirm", "confirmed", is_flag=True, help="Skip confirmation prompt")
def delete(page_id, confirmed):
    """Archive (delete) a page."""
    try:
        pid = _normalise_id(page_id)

        if not confirmed:
            from notion.modes import get_mode
            mode = get_mode()
            if mode == "interactive":
                if not click.confirm(f"Archive page {pid}?"):
                    click.echo("Cancelled.")
                    return

        client = NotionClient()
        result = client.patch(f"/pages/{pid}", {"archived": True})
        client.close()
        print_json({"status": "archived", "id": result.get("id")})

    except NotionCliError as e:
        handle_error(e)


@page.command("restore")
@click.argument("page_id")
def restore(page_id):
    """Restore an archived page."""
    try:
        pid = _normalise_id(page_id)
        client = NotionClient()
        result = client.patch(f"/pages/{pid}", {"archived": False})
        client.close()
        print_json({"status": "restored", "id": result.get("id")})

    except NotionCliError as e:
        handle_error(e)

File: src/commands/page.ts
import { Command } from 'commander'
import { createNotionClient, normaliseId, iteratePaginatedAPI } from '../client.js'
import { die, ExitCode } from '../errors.js'
import { printJSON, printId, printIds } from '../output.js'
import { parseKV, readDataInput, markdownToBlocks, looksLikeMarkdown } from '../coerce.js'
import { PropertyResolver } from '../schema.js'
import { confirm, getMode } from '../modes.js'

const resolver = new PropertyResolver()

export function registerPage(program: Command): void {
  const page = program.command('page').description('Page commands')

  // page list
  page.command('list <parent-id>')
    .description('List child pages under a page or block')
    .option('--output <format>', 'Output format: json|ids', 'json')
    .action(async (parentId: string, opts: { output: string }) => {
      const client = createNotionClient()
      const id = normaliseId(parentId)
      const results: unknown[] = []

      for await (const block of iteratePaginatedAPI(client.blocks.children.list, { block_id: id })) {
        if ((block as any).type === 'child_page') {
          results.push(block)
        }
      }

      if (opts.output === 'ids') {
        printIds(results as Array<{ id: string }>)
      } else {
        printJSON(results)
      }
    })

  // page create
  page.command('create <parent-id>')
    .description('Create a new page')
    .requiredOption('--title <title>', 'Page title')
    .option('--data <input>', 'Additional properties as JSON: inline, @file, or -')
    .option('--output <format>', 'Output format: json|id', 'json')
    .action(async (parentId: string, opts: { title: string; data?: string; output: string }) => {
      const client = createNotionClient()
      const id = normaliseId(parentId)

      let extraProps: Record<string, unknown> = {}
      if (opts.data) {
        extraProps = readDataInput(opts.data) as Record<string, unknown>
      }

      const page = await client.pages.create({
        parent: { page_id: id } as any,
        properties: {
          title: { title: [{ type: 'text', text: { content: opts.title } }] },
          ...extraProps,
        } as any,
      }) as any

      if (opts.output === 'id') {
        printId(page.id)
      } else {
        printJSON(page)
      }
    })

  // page get
  page.command('get <page-id>')
    .description('Retrieve a page')
    .option('--output <format>', 'Output format: json|properties', 'json')
    .action(async (pageId: string, opts: { output: string }) => {
      const client = createNotionClient()
      const p = await client.pages.retrieve({ page_id: normaliseId(pageId) }) as any

      if (opts.output === 'properties') {
        printJSON(p.properties ?? {})
      } else {
        printJSON(p)
      }
    })

  // page get-property
  page.command('get-property <page-id> <property-name>')
    .description('Retrieve a single property value')
    .action(async (pageId: string, propertyName: string) => {
      const client = createNotionClient()
      const id = normaliseId(pageId)
      const p = await client.pages.retrieve({ page_id: id }) as any
      const props = p.properties ?? {}

      // Case-insensitive property lookup
      const lower = propertyName.toLowerCase()
      const key = Object.keys(props).find(k => k.toLowerCase() === lower)
      if (!key) die(ExitCode.NOT_FOUND, `Property not found: ${propertyName}`)

      printJSON(props[key])
    })

  // page set
  page.command('set <page-id> [kv...]')
    .description('Update page properties')
    .option('--data <input>', 'JSON input: inline, @file, or -')
    .option('--no-cache', 'Skip schema cache')
    .action(async (pageId: string, kv: string[], opts: { data?: string; cache: boolean }) => {
      const client = createNotionClient()
      const id = normaliseId(pageId)

      let raw: Record<string, string | string[]> = parseKV(kv)
      if (opts.data) {
        const parsed = readDataInput(opts.data) as Record<string, string | string[]>
        raw = { ...raw, ...parsed }
      }

      // Try to get parent DB schema for type-aware coercion
      let properties: Record<string, unknown> = {}
      try {
        const p = await client.pages.retrieve({ page_id: id }) as any
        const dbId = p.parent?.database_id
        if (dbId) {
          const schema = await resolver.getSchema(dbId, client, { noCache: !opts.cache })
          properties = resolver.resolveAll(raw, schema)
        } else {
          // Page without DB parent — just set title/properties directly
          properties = raw as Record<string, unknown>
        }
      } catch {
        properties = raw as Record<string, unknown>
      }

      const updated = await client.pages.update({
        page_id: id,
        properties: properties as any,
      }) as any
      printJSON(updated)
    })

  // page append
  page.command('append <page-id>')
    .description('Append blocks to a page')
    .requiredOption('--data <input>', 'Markdown or JSON blocks: inline, @file, or -')
    .action(async (pageId: string, opts: { data: string }) => {
      const client = createNotionClient()
      const id = normaliseId(pageId)

      let blocks: unknown[]
      let rawInput: string

      if (opts.data.startsWith('@')) {
        const { readFileSync } = await import('fs')
        rawInput = readFileSync(opts.data.slice(1), 'utf-8')
      } else if (opts.data === '-') {
        const { readFileSync } = await import('fs')
        rawInput = readFileSync('/dev/stdin', 'utf-8')
      } else {
        rawInput = opts.data
      }

      if (looksLikeMarkdown(rawInput)) {
        blocks = markdownToBlocks(rawInput)
      } else {
        try {
          const parsed = JSON.parse(rawInput)
          blocks = Array.isArray(parsed) ? parsed : [parsed]
        } catch {
          blocks = markdownToBlocks(rawInput)
        }
      }

      const result = await client.blocks.children.append({
        block_id: id,
        children: blocks as any,
      }) as any
      printJSON(result)
    })

  // page delete
  page.command('delete <page-id>')
    .description('Archive (delete) a page')
    .option('--confirm', 'Skip confirmation prompt')
    .action(async (pageId: string, opts: { confirm?: boolean }) => {
      const mode = opts.confirm ? 'auto' : getMode()
      const ok = await confirm(`Archive page ${pageId}?`, mode)
      if (!ok) { console.log('Cancelled.'); return }
      const client = createNotionClient()
      const updated = await client.pages.update({
        page_id: normaliseId(pageId),
        archived: true,
      }) as any
      printJSON({ status: 'archived', id: updated.id })
    })

  // page restore
  page.command('restore <page-id>')
    .description('Restore an archived page')
    .action(async (pageId: string) => {
      const client = createNotionClient()
      const updated = await client.pages.update({
        page_id: normaliseId(pageId),
        archived: false,
      }) as any
      printJSON({ status: 'restored', id: updated.id })
    })

  // page get-markdown (v5 SDK exclusive — reads full page content as markdown)
  page.command('get-markdown <page-id>')
    .description('Retrieve a page\'s full content as markdown (v5 SDK)')
    .action(async (pageId: string) => {
      const client = createNotionClient()
      const result = await client.pages.retrieveMarkdown({ page_id: normaliseId(pageId) }) as any
      // Output raw markdown to stdout (not JSON-wrapped — useful for piping)
      process.stdout.write((result.markdown ?? result) + '\n')
    })

  // page set-markdown (v5 SDK exclusive — replaces full page content with markdown)
  page.command('set-markdown <page-id>')
    .description('Replace page content with markdown (v5 SDK)')
    .requiredOption('--data <input>', 'Markdown: inline string, @file, or -')
    .action(async (pageId: string, opts: { data: string }) => {
      const client = createNotionClient()
      const id = normaliseId(pageId)

      let markdown: string
      if (opts.data.startsWith('@')) {
        const { readFileSync } = await import('fs')
        markdown = readFileSync(opts.data.slice(1), 'utf-8')
      } else if (opts.data === '-') {
        const { readFileSync } = await import('fs')
        markdown = readFileSync('/dev/stdin', 'utf-8')
      } else {
        markdown = opts.data
      }

      const result = await client.pages.updateMarkdown({
        page_id: id,
        type: 'insert_content',
        insert_content: { content: markdown },
      }) as any
      printJSON(result)
    })

  // page move
  page.command('move <page-id> <new-parent-id>')
    .description('Move a page to a new parent')
    .action(async (pageId: string, newParentId: string) => {
      const client = createNotionClient()
      const id = normaliseId(pageId)
      const newParent = normaliseId(newParentId)

      const updated = await client.pages.move({
        page_id: id,
        parent: { page_id: newParent },
      }) as any
      printJSON({ status: 'moved', id: updated.id, new_parent: newParent })
    })
}
